### DISCUSSION TITLE
Reading Json from folder hierarchy

### DISCUSSION THREAD

**Discussion by uwegeercken** [duckdb/duckdb#7292] [Q&A]
Hello,

I  tried to import Json formatted data into duckdb. I have a rather complex Json file with many attributes and a lot of nesting. The Json files I have are in various folder in a hierarchy and one file contains one Json structure. The files are based on a Json schema and some of the attributes are optional.

Importing the files with read_json_auto worked for some files but on others it found errors (missing attributes). I wanted to increase the 'sample_size' parameter but it does not seem to work for Json files.

So I started writing the schema by hand. Which was quite some work! So I ended up with something like below (from the blog post about deeply nested json data):

```
CREATE TABLE test1 as
SELECT *
FROM read_json(
  '**/*.json',
  columns={
    .....
  },
  json_format='array_of_records',
  timestampformat='%Y-%m-%dT%H:%M:%SZ'
);
```
By the way it would be really nice if duckdb could use a Json schema definition file so one does not have to write the schema by hand.

This all worked and I imported some data. But now, if I want to add more data to duckdb (with the table already existing), I am stuck how to do that. If I use the copy syntax : COPY test1 FROM '**/*.json' it does not work. I think it does not support globs.

Can somebody help with this?

Regards,

Uwe

**Comment by lnkuiper** [comment]
Hi @uwegeercken the `sample_size` parameter increases the number of lines used to sample, but only up to a limit of 10 files. We have recently improved our `union_by_name` parameter for scanning files, but this has not yet been added to the JSON reader, but I plan to do this for the next release. This should help with sampling a little bit of each file to figure out the overall schema, and supplying `NULL` where attributes aren't found.

Globbing should work with JSON, but the `**` operator has only recently been added.

**Reply by uwegeercken** [reply]
Thank you for your answer. I am relatively new to duckdb so some of my attemtps may be premature... For the sample_size: I really interpreted that as "number of objects" and not "number of lines".

Potentionally I have several million of json files - all single files on S3 compatible storage (not AWS) - and I want to import them into duckdb. As indicated, the files have several hundred of fields, including some optional ones. I understand that I can store parts of the structure of the data as type JSON (and parse on read) to avoid to define all the fields that are there, but at the moment I wanted to define all of them.

So I tried to import a handful of files and duckdb generated a schema for me. Then I used more files until I hit an error and used the duckdb generated schema, amended it and recreated the table. But that is very cumbersome. I wish I could use the Json Schema I have for the files but that might not be that easy as it sounds because of the optional fields!?

An the other hand I understood that I can simply leave away parts of the structure and duckdb will simply ignore them during import. So I started defining the structure for the highest level in the json file and some of the substructures. So this is a way to evolve the required schema over time. But in the end it is still a lot of work.

Ideally I think it would be cool if duckdb could create the structure over all files it parses and create a unified schema. So if the first files do not have certain attrbitutes but later ones do, then a schema that covers them all is created. 

I will work on improving my know-how on duckdb and wait for the next release.
